### ISSUE TITLE
Bucket entry created in DB but missing from filesystem

### ISSUE THREAD

**Issue by akmithal** [noobaa/noobaa-core#6622]
### Environment info
- NooBaa Version: master-20210628
- Platform: OCP 4.6.16

### Actual behavior
1. bucket-142 was missing from filesystem when 200 buckets were created. It is present in DB 

### Expected behavior
1. No bucket should get missed out

### Steps to reproduce
1. Created 200 buckets in loop from bucket-1 to bucket-200. Got success message too for all. 
2. Found bucket-142 is missing from filesystem

```
[root@ocp-akshat-1-inf ~]# noobaa bucket status bucket-142
INFO[0000] ✅ Exists: NooBaa "noobaa"                    
INFO[0000] ✅ Exists: Service "noobaa-mgmt"              
INFO[0000] ✅ Exists: Secret "noobaa-operator"           
INFO[0000] ✅ Exists: Secret "noobaa-admin"              
INFO[0001] ✈️  RPC: bucket.read_bucket() Request: {Name:bucket-142} 
WARN[0001] RPC: GetConnection creating connection to wss://localhost:43767/rpc/ 0xc00007cf50 
INFO[0001] RPC: Connecting websocket (0xc00007cf50) &{RPC:0xc00036aff0 Address:wss://localhost:43767/rpc/ State:init WS:<nil> PendingRequests:map[] NextRequestID:0 Lock:{state:1 sema:0} ReconnectDelay:0s} 
INFO[0001] RPC: Connected websocket (0xc00007cf50) &{RPC:0xc00036aff0 Address:wss://localhost:43767/rpc/ State:init WS:<nil> PendingRequests:map[] NextRequestID:0 Lock:{state:1 sema:0} ReconnectDelay:0s} 
INFO[0001] ✅ RPC: bucket.read_bucket() Response OK: took 312.1ms 

Bucket status:
  Bucket                 : bucket-142
  Type                   : NAMESPACE
  Mode                   : OPTIMAL
  Num Objects            : N/A
  Data Size              : N/A
  Data Size Reduced      : N/A
  Data Space Avail       : N/A

```

```
sh-4.4# ls -lrt |grep bucket-14
drwxr-x---. 2 5000 800 4096 Jun 29 12:57 bucket-14
drwxr-x---. 2 5000 800 4096 Jun 29 13:04 bucket-140
drwxr-x---. 2 5000 800 4096 Jun 29 13:05 bucket-141
drwxr-x---. 2 5000 800 4096 Jun 29 13:05 bucket-143
drwxr-x---. 2 5000 800 4096 Jun 29 13:05 bucket-144
drwxr-x---. 2 5000 800 4096 Jun 29 13:05 bucket-145
drwxr-x---. 2 5000 800 4096 Jun 29 13:05 bucket-146
drwxr-x---. 2 5000 800 4096 Jun 29 13:05 bucket-147
drwxr-x---. 2 5000 800 4096 Jun 29 13:05 bucket-148
drwxr-x---. 2 5000 800 4096 Jun 29 13:06 bucket-149
sh-4.4# ls -lrt |grep bucket-142
sh-4.4# pwd
```
/nsfs/nsfs-nsr-1

### More information - Screenshots / Logs / Other output

Me and @jeniawhite witnessed this run together.

**Comment by akmithal** [comment]
[nbcore_postgres.gz](https://github.com/noobaa/noobaa-core/files/6738302/nbcore_postgres.gz)

**Comment by akmithal** [comment]
[noobaa_diagnostics_1624973152.tar.gz](https://github.com/noobaa/noobaa-core/files/6738308/noobaa_diagnostics_1624973152.tar.gz)

**Comment by akmithal** [comment]
Another similar instance but here bucket operation creation failed from app-server but bucket entry got created in DB. The bucket is not present in filesystem.


```
[root@ocp-akshat-1-inf test]# grep -R "bucket-1" .|grep ERROR
./noobaa-endpoint-7cb9d79574-2p257-endpoint.log:Jun-30 7:28:20.320 [Endpoint/8] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Resource>/bucket-1</Resource><RequestId>kqj5q6wk-fuq15n-p9f</RequestId></Error> PUT /bucket-1 {"accept-encoding":"identity","user-agent":"aws-cli/2.1.29 Python/3.8.8 Linux/4.18.0-305.3.1.el8_4.x86_64 exe/x86_64.rhel.8 prompt/off command/s3.mb","x-amz-date":"20210630T072820Z","x-amz-content-sha256":"61d056dc66f1882c0f4053be381523a7a28d384abde04fcf5b0021c716bb0ea1","authorization":"AWS4-HMAC-SHA256 Credential=BofHoptE4eAv8HFVV3SU/20210630/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=5d3a7b30e2da772559eb989deb74925787effde1a17419f0c8835b9694c10526","content-length":"154","host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-port":"443","x-forwarded-proto":"https","forwarded":"for=10.17.66.193;host=s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com;proto=https","x-forwarded-for":"10.17.66.193"} [RpcError: sorry, too many clients already] { rpc_code: 'INTERNAL', rpc_data: { retryable: true } }
./noobaa-endpoint-7cb9d79574-5p4q8-endpoint.log:Jun-30 7:28:37.127 [Endpoint/9] [ERROR] core.server.system_services.system_store:: SystemStoreData: buckets_by_name collision on key bucket-1 { _id: 60dc1d140dab1800234855c0, tag: '', name: SENSITIVE-f53e749d8f84e52b, system: <ref *1> {
./noobaa-endpoint-7cb9d79574-5p4q8-endpoint.log:Jun-30 7:28:37.160 [Endpoint/9] [ERROR] core.server.system_services.system_store:: SystemStoreData: buckets_by_name collision on key bucket-1 { _id: 60dc1d150dab1800234855c3, tag: '', name: SENSITIVE-f53e749d8f84e52b, system: <ref *1> {
./noobaa-endpoint-7cb9d79574-jn88l-endpoint.log:Jun-30 7:28:21.193 [Endpoint/9] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Resource>/bucket-1</Resource><RequestId>kqj5q7mi-er2z1f-y6f</RequestId></Error> PUT /bucket-1 {"accept-encoding":"identity","user-agent":"aws-cli/2.1.29 Python/3.8.8 Linux/4.18.0-305.3.1.el8_4.x86_64 exe/x86_64.rhel.8 prompt/off command/s3.mb","x-amz-date":"20210630T072821Z","x-amz-content-sha256":"61d056dc66f1882c0f4053be381523a7a28d384abde04fcf5b0021c716bb0ea1","authorization":"AWS4-HMAC-SHA256 Credential=BofHoptE4eAv8HFVV3SU/20210630/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=643fe005d56555ff1d72f7b683115d1b0c5b7012cc345dcf1682e5e780c7f30c","content-length":"154","host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-port":"443","x-forwarded-proto":"https","forwarded":"for=10.17.66.193;host=s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com;proto=https","x-forwarded-for":"10.17.66.193"} [RpcError: sorry, too many clients already] { rpc_code: 'INTERNAL', rpc_data: { retryable: true } }
./noobaa-endpoint-7cb9d79574-khl6q-endpoint.log:Jun-30 7:28:19.498 [Endpoint/8] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Resource>/bucket-1</Resource><RequestId>kqj5q69t-2az3bt-142s</RequestId></Error> PUT /bucket-1 {"accept-encoding":"identity","user-agent":"aws-cli/2.1.29 Python/3.8.8 Linux/4.18.0-305.3.1.el8_4.x86_64 exe/x86_64.rhel.8 prompt/off command/s3.mb","x-amz-date":"20210630T072819Z","x-amz-content-sha256":"61d056dc66f1882c0f4053be381523a7a28d384abde04fcf5b0021c716bb0ea1","authorization":"AWS4-HMAC-SHA256 Credential=BofHoptE4eAv8HFVV3SU/20210630/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=03b59c0a66dc521ea8319ec690571eed96923842324d49cbd8f4d920ac768c8a","content-length":"154","host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-port":"443","x-forwarded-proto":"https","forwarded":"for=10.17.66.193;host=s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com;proto=https","x-forwarded-for":"10.17.66.193"} [RpcError: remaining connection slots are reserved for non-replication superuser connections] { rpc_code: 'INTERNAL', rpc_data: { retryable: true } }
./noobaa-endpoint-7cb9d79574-kp89d-endpoint.log:Jun-30 7:28:37.124 [Endpoint/11] [ERROR] core.server.system_services.system_store:: SystemStoreData: buckets_by_name collision on key bucket-1 { _id: 60dc1d140dab1800234855c0, tag: '', name: SENSITIVE-f53e749d8f84e52b, system: <ref *1> {
./noobaa-endpoint-7cb9d79574-kp89d-endpoint.log:Jun-30 7:28:37.146 [Endpoint/11] [ERROR] core.server.system_services.system_store:: SystemStoreData: buckets_by_name collision on key bucket-1 { _id: 60dc1d150dab1800234855c3, tag: '', name: SENSITIVE-f53e749d8f84e52b, system: <ref *1> {
[root@ocp-akshat-1-inf test]# noobaa bucket list
INFO[0000] ✅ Exists: NooBaa "noobaa"                    
INFO[0000] ✅ Exists: Service "noobaa-mgmt"              
INFO[0000] ✅ Exists: Secret "noobaa-operator"           
INFO[0000] ✅ Exists: Secret "noobaa-admin"              
INFO[0001] ✈️  RPC: bucket.list_buckets() Request: <nil> 
WARN[0001] RPC: GetConnection creating connection to wss://localhost:37715/rpc/ 0xc000a2fae0 
INFO[0001] RPC: Connecting websocket (0xc000a2fae0) &{RPC:0xc00039d0e0 Address:wss://localhost:37715/rpc/ State:init WS:<nil> PendingRequests:map[] NextRequestID:0 Lock:{state:1 sema:0} ReconnectDelay:0s} 
INFO[0001] RPC: Connected websocket (0xc000a2fae0) &{RPC:0xc00039d0e0 Address:wss://localhost:37715/rpc/ State:init WS:<nil> PendingRequests:map[] NextRequestID:0 Lock:{state:1 sema:0} ReconnectDelay:0s} 
INFO[0001] ✅ RPC: bucket.list_buckets() Response OK: took 2.5ms 

BUCKET-NAME    
bucket-1       
first.bucket   

```

```
[root@hpo-11 akshat]# s3u1 mb s3://bucket-1
urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 's3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 's3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 's3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
make_bucket failed: s3://bucket-1 An error occurred (InternalError) when calling the CreateBucket operation (reached max retries: 2): We encountered an internal error. Please try again.
```

**Comment by akmithal** [comment]
@nimrod-becker , this is a DB corruption problem should be looked into it urgently I think.

This last run was done on master-20210629 build. I created 100 endpoints out of 39 came in Running state. First bucket creation failed there.

**Comment by akmithal** [comment]
[noobaa_diagnostics_1625038103.tar.gz](https://github.com/noobaa/noobaa-core/files/6738924/noobaa_diagnostics_1625038103.tar.gz)

**Comment by akmithal** [comment]
The views are different from app-server and from OCP

```
[root@hpo-11 ~]# s3u1 ls s3://
urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 's3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
2021-06-30 01:45:19 first.bucket
```


```
[root@ocp-akshat-1-inf test]# noobaa bucket status bucket-1
INFO[0000] ✅ Exists: NooBaa "noobaa"                    
INFO[0000] ✅ Exists: Service "noobaa-mgmt"              
INFO[0000] ✅ Exists: Secret "noobaa-operator"           
INFO[0000] ✅ Exists: Secret "noobaa-admin"              
INFO[0001] ✈️  RPC: bucket.read_bucket() Request: {Name:bucket-1} 
WARN[0001] RPC: GetConnection creating connection to wss://localhost:44729/rpc/ 0xc00097caf0 
INFO[0001] RPC: Connecting websocket (0xc00097caf0) &{RPC:0xc0004914f0 Address:wss://localhost:44729/rpc/ State:init WS:<nil> PendingRequests:map[] NextRequestID:0 Lock:{state:1 sema:0} ReconnectDelay:0s} 
INFO[0001] RPC: Connected websocket (0xc00097caf0) &{RPC:0xc0004914f0 Address:wss://localhost:44729/rpc/ State:init WS:<nil> PendingRequests:map[] NextRequestID:0 Lock:{state:1 sema:0} ReconnectDelay:0s} 
INFO[0001] ✅ RPC: bucket.read_bucket() Response OK: took 14.5ms 

Bucket status:
  Bucket                 : bucket-1
  Type                   : NAMESPACE
  Mode                   : OPTIMAL
  Num Objects            : N/A
  Data Size              : N/A
  Data Size Reduced      : N/A
  Data Space Avail       : N/A
```

**Comment by liranmauda** [comment]
Hi @akmithal 

I didn't manage to reproduce.
I am using master-20210630

From the endpoint pod: 
```
# ls -lr /nsfs/ns/ | grep bucket | wc -l
200
```

From the minikube:
```
$ ls -lr /nsfs/ | grep bucket | wc -l
200
```

From the db:
```
nbcore=# select jsonb_pretty(data -> 'name') from buckets;
  jsonb_pretty
----------------
 "bucket-188"
 "bucket-167"
 "bucket-166"
 "bucket-165"
 "bucket-142"
 "bucket-155"
 "bucket-148"
 "first.bucket"
 ...
 "bucket-45"
 "bucket-12"
 "bucket-8"
(201 rows)
```

**Comment by liranmauda** [comment]
Hi Akshat,
So far we have seen that on an overloaded cluster with many bucket requests, we are starting to pend on the RPC:
```
Jul-6 9:59:01.620 [Endpoint/9]    [L3] core.rpc.rpc:: RPC ROUTER default => wss://noobaa-mgmt.default.svc:443
Jul-6 9:59:01.620 [Endpoint/9]    [L2] core.rpc.rpc:: RPC _get_connection: existing address wss://noobaa-mgmt.default.svc:443 srv account_api.read_account_by_access_key connid wss://noobaa-mgmt.default.svc:443(jkkf1uo)
Jul-6 9:59:01.620 [Endpoint/9]    [L1] core.rpc.rpc:: RPC._request: START srv account_api.read_account_by_access_key reqid 11@wss://noobaa-mgmt.default.svc:443(jkkf1uo) connid wss://noobaa-mgmt.default.svc:443(jkkf1uo)
Jul-6 9:59:01.621 [Endpoint/9]    [L1] core.rpc.rpc:: RPC._request: SEND srv account_api.read_account_by_access_key reqid 11@wss://noobaa-mgmt.default.svc:443(jkkf1uo) connid wss://noobaa-mgmt.default.svc:443(jkkf1uo)
Jul-6 9:59:01.621 [Endpoint/9]    [L1] core.rpc.rpc:: RPC._request: WAIT srv account_api.read_account_by_access_key reqid 11@wss://noobaa-mgmt.default.svc:443(jkkf1uo) connid wss://noobaa-mgmt.default.svc:443(jkkf1uo)
Jul-6 9:59:15.225 [Endpoint/9]    [L4] core.rpc.rpc:: RPC PING wss://noobaa-mgmt.default.svc:443(jkkf1uo)
```
WAIT take ~14 sec

Rebuild takes a lot of time:
```
Jul-6 9:01:52.223 [Endpoint/88]    [L1] core.server.system_services.system_store:: SystemStore: rebuild took 8205.7ms
```
We have pending items in the Semafore:
```
Jul-6 10:06:51.527 [Endpoint/9]    [L2] core.util.semaphore:: Semaphore release updating value  9  ->  10

```
And the Core CPU is maxed out (My system have only 100m CPU):
```
NAME                                               CPU(cores)   MEMORY(bytes)
noobaa-core-0                                      99m          734Mi
noobaa-db-pg-0                                     5m           76Mi
noobaa-default-backing-store-noobaa-pod-3ef1e66b   3m           57Mi
noobaa-endpoint-555554b56c-6pfsp                   12m          77Mi
noobaa-operator-bbbf64cf6-7trdr                    2m           31Mi
```

**Comment by nimrod-becker** [comment]
With these two fixes 
https://github.com/noobaa/noobaa-core/pull/6657
https://github.com/noobaa/noobaa-core/pull/6654

Is it still happening? We think this might be related

**Comment by akmithal** [comment]
This defect seems to be fixed in master-20210802. Closing this bug now.
